polaris/hub/storage.py

```diff
@@ -0,0 +1,431 @@
+                    ContentMD5=b64encode(md5_hash.digest()).decode(),
+                    Metadata={
+                        "md5sum": md5_hash.hexdigest(),
+                    },
+                )
+
+    def __delitem__(self, key: str) -> None:
+        """
+        Removing a key from the store is not supported.
+        """
+        raise NotImplementedError(f'{type(self)} is not erasable, cannot call "del store[key]"')
+
+    def __contains__(self, key: str) -> bool:
+        """
+        Checks the existence of a key in the store.
+
+        If the intent is to download the value after this check, it is more efficient to
+        attempt tp retrieve it and handle the KeyError from a non-existent key.
+        """
+        with handle_s3_errors():
+            try:
+                self.s3_client.head_object(Bucket=self.bucket_name, Key=self._full_key(key))
+                return True
+            except self.s3_client.exceptions.NoSuchKey:
+                return False
+            except ClientError as e:
+                if e.response["Error"]["Code"] == "404":
+                    return False
+                raise e
+
+    def __iter__(self) -> Sequence[str]:
+        """
+        Iterate through all the keys in the store.
+        """
+        with handle_s3_errors():
+            # `list_objects_v2` returns a max of 1000 keys per request, so paginate requests
+            paginator = self.s3_client.get_paginator("list_objects_v2")
+            page_iterator = paginator.paginate(Bucket=self.bucket_name, Prefix=self.prefix)
+
+            for page in page_iterator:
+                for obj in page.get("Contents", []):
+                    yield obj["Key"][len(self.prefix) + 1 :]
+
+    def __len__(self) -> int:
+        """
+        Number of keys in the store.
+        """
+        with handle_s3_errors():
+            # `list_objects_v2` returns a max of 1000 keys per request, so paginate requests
+            paginator = self.s3_client.get_paginator("list_objects_v2")
+            page_iterator = paginator.paginate(Bucket=self.bucket_name, Prefix=self.prefix)
+
+            return sum((page["KeyCount"] for page in page_iterator))
+
+
+class StorageTokenAuth:
+    token: HubStorageOAuth2Token | None
+
+    def __init__(self, token: dict[str, Any] | None, *args):
+        self.token = None
+        if token:
+            self.set_token(token)
+
+    def set_token(self, token: dict[str, Any] | HubStorageOAuth2Token):
+        self.token = HubStorageOAuth2Token(**token) if isinstance(token, dict) else token
+
+
+class StorageSession(OAuth2Client):
+    """
+    A context manager for managing a storage session, with token exchange and token refresh capabilities.
+    Each session is associated with a specific scope and resource.
+    """
+
+    polaris_protocol = "polarisfs"
+
+    token_auth_class = StorageTokenAuth
+
+    def __init__(self, hub_client, scope: Scope, resource: str):
```

Comment:
I'd love to type the `hub_client` param, but that would cause a circular import problem. One more reason we should refactor the client's structure.

I've added a type for artifact URNs, so I'll use this in the signature.